### ISSUE TITLE
static tokens file not working

### ISSUE THREAD

**Issue by mengesb** [rundeck/rundeck#4536]
**Describe the bug**
I've defined a static API token in the static tokens file and I'm unable to use that token.

**My Rundeck detail**

* Rundeck version: 3.0.13-20190123
* install type: rpm
* OS Name/version: RHEL 7.x
* DB Type/version: H2

**To Reproduce**
1. Generate `/etc/rundeck/tokens.properties`
2. Update `/etc/rundeck/framework.properties` and set `rundeck.tokens.file=/etc/rundeck/tokens.properties`
3. Restart rundeckd
4. Attempt `curl -vk https://host.domain.tld/api/14/system/info\?authtoken\=__________` using the static token

**Expected behavior**
API call works correctly

**Screenshots**
```
[root@rd ~]# grep tokens /etc/rundeck/framework.properties 
rundeck.tokens.file=/etc/rundeck/tokens.properties
[root@rd ~]# cat /etc/rundeck/tokens.properties
# Generated by Chef for rd
# Local modifications will be overwritten.

admin: BFrVx****

# ---
curl -vk https://host.domain.tld/api/14/system/info\?authtoken\=BFrVx****
# ---
2019-02-23 06:07:44.699 ERROR --- [qtp490625541-26] rundeck.interceptors.SetUserInterceptor  : Invalid API token used: BFrVx****
2019-02-23 06:07:44.701 ERROR --- [qtp490625541-26] r.interceptors.AuthorizationInterceptor  : (Token:BFrVx****) UNAUTHORIZED for api/apiSystemInfo
```

*Desktop (please complete the following information):**
 - OS: N/A
 - Browser N/A
 - Version N/A

**Additional context**
I've sanitized the above.... yes the API token is the correct length. I've also tried with the newer format with the `username: token,role1,role2,rol3` method but this isn't working either

If I use an API token created via the UI, things work as expected. I've attempted several reboots like #971 suggested and #4124 I read but wasn't any help. I still can't use a token inserted in the tokens file.

**Comment by mengesb** [comment]
File permissions should be OK:
```
-rw-r-----.  1 rundeck rundeck 3033 Feb 23 06:00 framework.properties
-rw-r-----.  1 rundeck rundeck  606 Feb 10 21:50 rundeck-config.properties
-rw-r-----.  1 rundeck rundeck  137 Feb 23 06:16 tokens.properties
```

**Comment by mengesb** [comment]
In reference to the other two issues cited; exporting a `RD_TOKEN` variable has no effect nor does restarting the service.

**Comment by sjrd218** [comment]
@mengesb I cannot reproduce the issue you are reporting. It seems your tokens.properties file might not be getting read. 

Can you put a `rundeck-config.groovy` file in /etc/rundeck with the following contents:

    import org.slf4j.LoggerFactory
    import ch.qos.logback.classic.LoggerContext
    import ch.qos.logback.classic.Level

    LoggerContext loggerContext = (LoggerContext)LoggerFactory.getILoggerFactory();
    loggerContext.getLogger("rundeckapp.BootStrap").setLevel(Level.DEBUG)

restart Rundeck, then post the part of your service.log that shows the bootstrap information.

**Comment by mengesb** [comment]
@sjrd218 

```text
2019-02-28 15:22:28.424  INFO --- [           main] rundeckapp.BootStrap                     : Starting Rundeck 3.0.13-20190123 (2019-01-24) ...
2019-02-28 15:22:28.431  INFO --- [           main] rundeckapp.BootStrap                     : using rdeck.base config property: /var/lib/rundeck
2019-02-28 15:22:28.449  INFO --- [           main] rundeckapp.BootStrap                     : loaded configuration: /etc/rundeck/framework.properties
2019-02-28 15:22:28.460 DEBUG --- [           main] rundeckapp.BootStrap                     : Loaded null tokens from tokens file: /etc/rundeck/tokens.properties...
2019-02-28 15:22:28.488 DEBUG --- [           main] rundeckapp.BootStrap                     : BEGIN: Initialized WorkflowService
2019-02-28 15:22:28.493 DEBUG --- [           main] rundeckapp.BootStrap                     : Initialized WorkflowService in 2ms
2019-02-28 15:22:28.495 DEBUG --- [           main] rundeckapp.BootStrap                     : BEGIN: Initialized ScmService
2019-02-28 15:22:28.497 DEBUG --- [           main] rundeckapp.BootStrap                     : Initialized ScmService in 2ms
2019-02-28 15:22:28.497  INFO --- [           main] rundeckapp.BootStrap                     : RSS feeds disabled
2019-02-28 15:22:28.497  INFO --- [           main] rundeckapp.BootStrap                     : Using jaas authentication
2019-02-28 15:22:28.503  INFO --- [           main] rundeckapp.BootStrap                     : Preauthentication is disabled
2019-02-28 15:22:28.546  INFO --- [           main] rundeckapp.BootStrap                     : Rundeck is ACTIVE: executions can be run.
2019-02-28 15:22:28.562 DEBUG --- [           main] rundeckapp.BootStrap                     : executionService.cleanupRunningJobs: starting asynchronously
2019-02-28 15:22:28.731 DEBUG --- [           main] rundeckapp.BootStrap                     : scheduledExecutionService.rescheduleJobs: starting asynchronously
2019-02-28 15:22:28.741 DEBUG --- [           main] rundeckapp.BootStrap                     : logFileStorageService.resumeIncompleteLogStorage: resuming asynchronously
2019-02-28 15:22:28.751  WARN --- [           main] rundeckapp.BootStrap                     : [Development Mode] Usage of H2 database is recommended only for development and testing
2019-02-28 15:22:28.756  INFO --- [           main] rundeckapp.BootStrap                     : Rundeck startup finished in 485ms
Grails application running at http://127.0.0.1:4440 in environment: production```

**Comment by mengesb** [comment]
so what would be the reason it says `Loaded null tokens from tokens file: /etc/rundeck/tokens.properties` ?

**Comment by sjrd218** [comment]
@mengesb The "Loaded null tokens..." is a bug in the debug message. Ironically, that message means your tokens file is being loaded correctly. 

Can you try modifying your curl command to send the token as a header: `curl -vk http://127.0.0.1:4440/api/29/projects -H "X-Rundeck-Auth-Token: 123YourApiToken456"` 

Note: the `tokens.properties` file will not support multiple tokens per user. So if you have multiple entries like: 
```
admin: 123abc
admin: abc123
``` 
only the last token will be valid.

**Comment by mengesb** [comment]
There's only one token in that file, and only for admin.

```
$ curl -H "X-Rundeck-Auth-Token: BFrVx****" -vk https://host.domain.tld/api/14/system/info
...
* Expire in 50 ms for 1 (transfer 0x7f8b57805400)
* Expire in 50 ms for 1 (transfer 0x7f8b57805400)
*   Trying 172.21.4.134...
* TCP_NODELAY set
* Expire in 200 ms for 4 (transfer 0x7f8b57805400)
* Connected to host.domain.tld (172.21.4.134) port 443 (#0)
* ALPN, offering http/1.1
* WARNING: disabling hostname validation also disables SNI.
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
* Server certificate: *.domain.tld
* Server certificate: Thawte TLS RSA CA G1
* Server certificate: DigiCert Global Root G2
> GET /api/14/system/info HTTP/1.1
> Host: host.domain.tld
> User-Agent: curl/7.64.0
> Accept: */*
> X-Rundeck-Auth-Token: BFrVx****
>
< HTTP/1.1 403 Forbidden
< Server: nginx
< Date: Thu, 28 Feb 2019 15:59:13 GMT
< Content-Type: text/xml;charset=utf-8
< Transfer-Encoding: chunked
< Connection: keep-alive
< X-Frame-Options: deny
< X-XSS-Protection: 1
< X-Content-Type-Options: nosniff
< Content-Security-Policy: default-src 'none' ; script-src 'self' 'unsafe-inline' 'unsafe-eval' ; style-src 'self' 'unsafe-inline' ; img-src 'self' https://media.rundeck.org ; font-src 'self' data: ; connect-src 'self' ; form-action 'self' ;
< X-Application-Context: application:production:4440
< Set-Cookie: JSESSIONID=node043wg4dyf3tvq1bpebh5oe3sr00.node0;Path=/;Secure;HttpOnly
< Expires: Thu, 01 Jan 1970 00:00:00 GMT
< Strict-Transport-Security: max-age=31536000; includeSubDomains
<
* Connection #0 to host us1a-rundeck1-dbz.anaplan.io left intact
<result error='true' apiversion='29'><error code='unauthorized'><message>(Token:BFrVx****) is not authorized for: /api/14/system/info</message></error></result>%
```

Real token not pasted for security reasons.

I get the same results going directly to the rundeckd port on 4440 from the local host as I do through NGiNX above

**Comment by mengesb** [comment]
Any thoughts @sjrd218 ?

**Comment by mengesb** [comment]
tokens.properties:
```text
# Generated by Chef for host.domain.tld
# Local modifications will be overwritten.

admin: BFrVx****
```

**Comment by mengesb** [comment]
@sjrd218 / @gschueler ; So per the PR merged for 3.0.17, I've re-deployed the instance, and also put the logging in as referenced above. It did show that it loaded 1 token from the file... and READ permissions were finally granted, however it seems write permissions were not granted.

I updated my tokens file to specify the group as well and I was finally able to run my rundeck automation from terraform with the provided static key. Not sure what the root of the problem was in 3.0.13, but something was resolved between then and now.
